File: Graph/2668번 - 숫자고르기/2668.py
import sys

input = sys.stdin.readline

n = int(input())
arr = [0] + [int(input()) for _ in range(n)]
res = []

def recursive(st, val, visited):
    if visited[val]:
        if st == val:
            res.append(st)
        return
    visited[val] = True
    recursive(st, arr[val], visited)

def solve():
    for i in range(1, n + 1):
        vis = [False for _ in range(n + 1)]
        vis[i] = True
        recursive(i, arr[i], vis)
    
    print(len(res))
    for r in res:
        print(r)

if __name__ == '__main__':
    solve()

File: Graph/2668번 - 숫자고르기/2668.cpp
#include <iostream>
#include <vector>

using namespace std;

int n;
int arr[2][101];
vector<int> res;

void input() {
    cin >> n;
    for(int i = 1; i <= n; i++) {
        cin >> arr[1][i];
    }
}

void recursive(int st, int val) {
    if(arr[0][val]) {
        if(st == val) res.push_back(st);
        return;
    }
    arr[0][val] = 1;
    recursive(st, arr[1][val]);
}

void solve() {
    for(int i = 1; i <= n; i++) {
        fill_n(&arr[0][0], 101, 0);
        arr[0][i] = 1;
        recursive(i, arr[1][i]);
    }
    cout << res.size() << '\n';
    for(int &it : res) cout << it << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}